s3tests_boto3/functional/test_s3select.py

```diff
@@ -0,0 +1,410 @@
+    get_alt_client,
+    get_tenant_client,
+    get_tenant_iam_client,
+    get_tenant_user_id,
+    get_buckets_list,
+    get_objects_list,
+    get_main_kms_keyid,
+    get_secondary_kms_keyid,
+    nuke_prefixed_buckets,
+    )
+
+import boto
+import boto.s3.connection
+import sys
+import random
+from botocore.client import Config
+
+region_name = ''
+
+def get_connection():
+    conn = boto.connect_s3(
+        aws_access_key_id = get_main_aws_access_key(),
+        aws_secret_access_key = get_main_aws_secret_key(),
+        host = get_config_host(),
+        port = get_config_port(),
+        is_secure=False,               # uncomment if you are not using ssl
+        calling_format = boto.s3.connection.OrdinaryCallingFormat(),
+        )
+
+    return conn
+
+def create_csv_object_for_datetime(rows,columns):
+        result = ""
+        for i in range(rows):
+            row = "";
+            for y in range(columns):
+                row = row + "{}{:02d}{:02d}-{:02d}{:02d}{:02d},".format(random.randint(0,100)+1900,random.randint(1,12),random.randint(1,28),random.randint(0,23),random.randint(0,59),random.randint(0,59),);
+            result += row + "\n"
+
+        return result
+
+def create_random_csv_object(rows,columns,col_delim=",",record_delim="\n",csv_schema=""):
+        result = ""
+        if len(csv_schema)>0 :
+            result = csv_schema + record_delim
+
+        for i in range(rows):
+            row = "";
+            for y in range(columns):
+                row = row + "{}{}".format(random.randint(0,1000),col_delim);
+            result += row + record_delim
+
+        return result
+
+
+def upload_csv_object(bucket_name,new_key,obj):
+        conn = get_connection()
+        conn.create_bucket( bucket_name )
+        bucket = conn.get_bucket( bucket_name )
+
+        k1 = bucket.new_key( new_key );
+        k1.set_contents_from_string( obj );
+        
+    
+def run_s3select(bucket,key,query,column_delim=",",row_delim="\n",quot_char='"',esc_char='\\',csv_header_info="NONE"):
+
+    s3 = boto3.client('s3',#'sns',
+        endpoint_url=get_config_endpoint(),
+        aws_access_key_id=get_main_aws_access_key(),
+        region_name=region_name,
+        aws_secret_access_key=get_main_aws_secret_key())
+        #config=Config(signature_version='v2'))
+
+
+    r = s3.select_object_content(
+        Bucket=bucket,
+        Key=key,
+        ExpressionType='SQL',
+        InputSerialization = {"CSV": {"RecordDelimiter" : row_delim, "FieldDelimiter" : column_delim,"QuoteEscapeCharacter": esc_char, "QuoteCharacter": quot_char, "FileHeaderInfo": csv_header_info}, "CompressionType": "NONE"},
+        OutputSerialization = {"CSV": {}},
+        Expression=query,)
+    
+    result = ""
+    for event in r['Payload']:
+        if 'Records' in event:
+            records = event['Records']['Payload'].decode('utf-8')
+            result += records
+    
+    return result
+
+def remove_xml_tags_from_result(obj):
```

Comment:
why not run the result through a standard xml parser?
this would also validate that the xml format is correct, and what would any client implementation do

Reply:
upon output serialization is implemented, maybe it's better to run standard xml-parser.
currently the /s3.select_object_content/ validates correct output.

Reply:
where should "output serialization" be implemented?
is this in the client or in the product?
my suggestion here is to run the output via an xml parser, to make sure it is correct

Reply:
"output serialization" implemented for the AWS cli (here its boto3, same).
its part of spec implemented in RGW (header-type,message-type,crc,and xml tags).

Reply:
there is no need to validate the output ourselves if the boto3 client already does that
e.g. if we reply invalid xml tags or structure the boto3 client would error?

Reply:
yes.
in the course of output serialization implementation, it reject RGW return response (after accepting request and processing it), unless it comply to client(AWS) specification.

did not try to validate response with xml-parser.
the other important clients would be analytical utilities such as presto.